cardano-db-sync/src/Cardano/DbSync/Cache.hs

```diff
@@ -123,75 +125,85 @@ insertStakeAddress rewardAddr stakeCredBs =
-queryStakeAddrWithCacheRetBs cacheStatus cacheUA nw cred = do
-  case cacheStatus of
+queryStakeAddrWithCacheRetBs trce cache cacheUA nw cred = do
+  let !bs = Ledger.serialiseRewardAccount (Ledger.RewardAccount nw cred)
+  case cache of
     NoCache -> do
-      let !bs = Ledger.serialiseRewardAccount (Ledger.RewardAccount nw cred)
       mapLeft (,bs) <$> queryStakeAddress bs
     ActiveCache ci -> do
-      mp <- liftIO $ readTVarIO (cStakeCreds ci)
-      (mAddrId, mp') <- queryStakeAddrAux cacheUA mp (cStats ci) nw cred
-      liftIO $ atomically $ writeTVar (cStakeCreds ci) mp'
-      pure mAddrId
+      prevCache <- liftIO $ readTVarIO (cStakeRawHashes ci)
+      let isNewCache = LRU.getSize prevCache < 1
+      -- populate from db if the cache is empty
+      currentCache <-
+        if isNewCache
+          then do
+            liftIO $ logInfo trce "Stake Raw Hashes cache is new and empty. Populating with addresses from db..."
+            queryRes <- DB.queryAddressWithReward (fromIntegral $ LRU.getCapacity prevCache)
+            liftIO $ atomically $ writeTVar (cStakeRawHashes ci) $ LRU.fromList queryRes prevCache
+            liftIO $ logInfo trce "Population of cache complete."
+            liftIO $ readTVarIO (cStakeRawHashes ci)
+          else pure prevCache
 
-queryStakeAddrAux ::
-  MonadIO m =>
-  CacheUpdateAction ->
-  StakeAddrCache ->
-  StrictTVar IO CacheStatistics ->
-  Network ->
-  StakeCred ->
-  ReaderT SqlBackend m (Either (DB.LookupFail, ByteString) DB.StakeAddressId, StakeAddrCache)
-queryStakeAddrAux cacheUA mp sts nw cred =
-  case Map.lookup cred mp of
-    Just addrId -> do
-      liftIO $ hitCreds sts
-      case cacheUA of
-        EvictAndUpdateCache -> pure (Right addrId, Map.delete cred mp)
-        _ -> pure (Right addrId, mp)
-    Nothing -> do
-      liftIO $ missCreds sts
-      let !bs = Ledger.serialiseRewardAccount (Ledger.RewardAccount nw cred)
-      mAddrId <- mapLeft (,bs) <$> queryStakeAddress bs
-      case (mAddrId, cacheUA) of
-        (Right addrId, UpdateCache) -> pure (Right addrId, Map.insert cred addrId mp)
-        (Right addrId, _) -> pure (Right addrId, mp)
-        (err, _) -> pure (err, mp)
+      case LRU.lookup bs currentCache of
+        Just (addrId, lruCache) -> do
+          liftIO $ hitCreds (cStats ci)
+          case cacheUA of
+            EvictAndUpdateCache -> do
+              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) $ LRU.delete bs lruCache
+              pure $ Right addrId
+            _other -> do
+              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) lruCache
+              pure $ Right addrId
+        Nothing -> do
+          queryRes <- mapLeft (,bs) <$> queryStakeAddress bs
+          liftIO $ missCreds (cStats ci)
+          case queryRes of
+            Left _ -> pure queryRes
+            Right stakeAddrsId -> do
+              when (isCacheActive cache) $
```

Comment:
This code only runs if the cache is Active. We actually want to check that the specific request wants to update the cache

> I think we should update the cache only if UpdateCache is set.

Reply:
sorry misunderstood 🤦‍♀️